File: src/main/java/xyz/chunkstories/api/converter/MinecraftBlocksTranslator.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.converter;

import java.io.BufferedReader;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;

import xyz.chunkstories.api.content.Asset;
import xyz.chunkstories.api.converter.mappings.Mapper;
import xyz.chunkstories.api.voxel.Voxel;
import xyz.chunkstories.api.world.GameInstance;
import xyz.chunkstories.api.world.cell.FutureCell;

//TODO kotlinize
/** Maps minecraft ids to chunkstories's */
public class MinecraftBlocksTranslator {
	public final int MINECRAFT_IDS_CAP = 256;
	public final int MINECRAFT_METADATA_SIZE = 16;

	private Mapper[] mappers = new Mapper[MINECRAFT_IDS_CAP * MINECRAFT_METADATA_SIZE];

	public GameInstance getGameInstance() {
		return gameInstance;
	}

	enum Section
	{
		NONE, MAPPERS, MAPPINGS,
	}

	private GameInstance gameInstance;

	public MinecraftBlocksTranslator(GameInstance gameInstance, Asset asset) throws IOException {
		this.gameInstance = gameInstance;
		BufferedReader reader = new BufferedReader(asset.reader());

		Section currentSection = Section.MAPPINGS;
		Map<String, Constructor<? extends Mapper>> customMappers = new HashMap<>();

		String line;
		while ((line = reader.readLine()) != null) {
			if (!line.startsWith("#") && line.length() > 0) {

				if (line.startsWith("_")) {
					String sectionName = line.substring(1);
					if (sectionName.equalsIgnoreCase("mappers"))
						currentSection = Section.MAPPERS;
					else if (sectionName.equalsIgnoreCase("mappings"))
						currentSection = Section.MAPPINGS;

					continue;
				}

				if (currentSection == Section.MAPPERS) {
					String[] splitted = line.split(" ");
					if (splitted.length == 2) {
						String name = splitted[0];
						String className = splitted[1];

						try {
							@SuppressWarnings("unchecked")
							Class<? extends Mapper> mapperClass = (Class<? extends Mapper>) gameInstance.getContent().getModsManager().getClassByName(className);
							Constructor<? extends Mapper> mapperConstructor = mapperClass.getConstructor(Voxel.class);
							customMappers.put(name, mapperConstructor);
						} catch (Exception e) {
							System.out.println(e);
						}
					}
				} else if (currentSection == Section.MAPPINGS) {
					String[] splitted = line.split(" ");
					if (splitted.length >= 2) {
						String mc = splitted[0];
						String cs = splitted[1];

						String special = splitted.length >= 3 ? splitted[2] : null;

						int minecraftID;
						byte minecraftMeta = -1;

						// Read the minecraft thingie
						if (mc.contains(":")) {
							minecraftID = Integer.parseInt(mc.split(":")[0]);
							minecraftMeta = Byte.parseByte(mc.split(":")[1]);
						} else {
							minecraftID = Integer.parseInt(mc);
						}

						// Read the cs part
						String chunkStoriesName = null;
						int chunkStoriesMeta = 0;
						if (cs.contains(":")) {
							chunkStoriesName = cs.split(":")[0];
							chunkStoriesMeta = Integer.parseInt(cs.split(":")[1]);
						} else {
							chunkStoriesName = cs;
						}

						Voxel voxel = gameInstance.getContent().getVoxels().getVoxel(chunkStoriesName);
						if (voxel == null) {
							System.out.println("Error: Voxel '" + chunkStoriesName + "' is nowhere to be found in the loaded content.");
							System.out.println("Skipping line : '" + line + "'.");
							continue;
						}

						Mapper mapper;
						if (special == null) {
							mapper = new TrivialMapper(voxel, chunkStoriesMeta);
						} else {

							Constructor<? extends Mapper> mapperConstructor = customMappers.get(special);

							if (mapperConstructor == null) {
								System.out.println("Error: mapper '" + special + "' was not recognised.");
								System.out.println("Skipping line : '" + line + "'.");
								continue;
							}

							try {
								mapper = mapperConstructor.newInstance(voxel);
							} catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
								e.printStackTrace();
								continue;
							}
						}

						// Fill the relevant cases
						if (minecraftMeta == -1) {
							for (int i = 0; i < 16; i++) {
								mappers[minecraftID * 16 + i] = mapper;
							}
						} else {
							mappers[minecraftID * 16 + minecraftMeta] = mapper;
						}
					}
				}
			}
		}

		reader.close();
	}

	public Mapper getMapper(int minecraftID, byte minecraftMeta) {
		minecraftID &= 0xFF;
		minecraftMeta &= 0xF;

		return mappers[minecraftID * 16 + minecraftMeta];
	}

	class TrivialMapper extends Mapper {

		int meta;

		TrivialMapper(Voxel voxel, int meta) {
			super(voxel);
			this.meta = meta;
		}

		@Override
		public void output(int minecraftId, byte minecraftMeta, FutureCell future) {
			future.setVoxel(voxel);
			future.setMetaData(meta);
		}
	}
}

File: src/main/java/xyz/chunkstories/api/converter/MinecraftBlocksTranslator.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.api.converter

import xyz.chunkstories.api.world.GameInstance
import xyz.chunkstories.api.content.Asset
import xyz.chunkstories.api.block.BlockType
import xyz.chunkstories.api.world.cell.MutableCellData
import java.io.BufferedReader
import java.util.HashMap
import java.lang.Exception
import java.lang.InstantiationException
import java.lang.IllegalAccessException
import java.lang.IllegalArgumentException
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException

// TODO belongs in implementation
class MinecraftBlocksTranslator(val gameInstance: GameInstance, asset: Asset) {
    val MINECRAFT_IDS_CAP = 256
    val MINECRAFT_METADATA_SIZE = 16
    private val mappers = arrayOfNulls<Mapper>(MINECRAFT_IDS_CAP * MINECRAFT_METADATA_SIZE)

    internal enum class Section {
        NONE, MAPPERS, MAPPINGS
    }

    fun getMapper(minecraftID: Int, minecraftMeta: Int): Mapper? {
        return mappers[(minecraftID and 0xFF) * 16 + (minecraftMeta and 0xF)]
    }

    internal inner class TrivialMapper(voxel: BlockType, var metadata: Int) : Mapper(voxel) {
        override fun output(minecraftId: Int, minecraftMeta: Byte, output: MutableCellData) {
            output.blockType = blockType
            output.extraData = metadata
        }
    }

    init {
        val reader = BufferedReader(asset.reader())
        var currentSection = Section.MAPPINGS
        val customMappers: MutableMap<String, Constructor<out Mapper>> = HashMap()
        var line: String
        while (reader.readLine().also { line = it } != null) {
            if (!line.startsWith("#") && line.isNotEmpty()) {
                if (line.startsWith("_")) {
                    val sectionName = line.substring(1)
                    if (sectionName.equals("mappers", ignoreCase = true)) currentSection = Section.MAPPERS else if (sectionName.equals("mappings", ignoreCase = true)) currentSection = Section.MAPPINGS
                    continue
                }
                if (currentSection == Section.MAPPERS) {
                    val splitted = line.split(" ").toTypedArray()
                    if (splitted.size == 2) {
                        val name = splitted[0]
                        val className = splitted[1]
                        try {
                            val mapperClass = gameInstance.content.modsManager.getClassByName(className) as Class<out Mapper>?
                            val mapperConstructor = mapperClass!!.getConstructor(BlockType::class.java)
                            customMappers[name] = mapperConstructor
                        } catch (e: Exception) {
                            println(e)
                        }
                    }
                } else if (currentSection == Section.MAPPINGS) {
                    val splitted = line.split(" ").toTypedArray()
                    if (splitted.size >= 2) {
                        val mc = splitted[0]
                        val cs = splitted[1]
                        val special = if (splitted.size >= 3) splitted[2] else null
                        var minecraftID: Int
                        var minecraftMeta: Byte = -1

                        // Read the minecraft thingie
                        if (mc.contains(":")) {
                            minecraftID = mc.split(":").toTypedArray()[0].toInt()
                            minecraftMeta = mc.split(":").toTypedArray()[1].toByte()
                        } else {
                            minecraftID = mc.toInt()
                        }

                        // Read the cs part
                        var chunkStoriesName: String
                        var chunkStoriesMeta = 0
                        if (cs.contains(":")) {
                            chunkStoriesName = cs.split(":").toTypedArray()[0]
                            chunkStoriesMeta = cs.split(":").toTypedArray()[1].toInt()
                        } else {
                            chunkStoriesName = cs
                        }
                        val voxel = gameInstance.content.blockTypes[chunkStoriesName]
                        if (voxel == null) {
                            println("Error: Voxel '$chunkStoriesName' is nowhere to be found in the loaded content.")
                            println("Skipping line : '$line'.")
                            continue
                        }
                        var mapper: Mapper
                        mapper = if (special == null) {
                            TrivialMapper(voxel, chunkStoriesMeta)
                        } else {
                            val mapperConstructor = customMappers[special]
                            if (mapperConstructor == null) {
                                println("Error: mapper '$special' was not recognised.")
                                println("Skipping line : '$line'.")
                                continue
                            }
                            try {
                                mapperConstructor.newInstance(voxel)
                            } catch (e: InstantiationException) {
                                e.printStackTrace()
                                continue
                            } catch (e: IllegalAccessException) {
                                e.printStackTrace()
                                continue
                            } catch (e: IllegalArgumentException) {
                                e.printStackTrace()
                                continue
                            } catch (e: InvocationTargetException) {
                                e.printStackTrace()
                                continue
                            }
                        }

                        // Fill the relevant cases
                        if (minecraftMeta.toInt() == -1) {
                            for (i in 0..15) {
                                mappers[minecraftID * 16 + i] = mapper
                            }
                        } else {
                            mappers[minecraftID * 16 + minecraftMeta] = mapper
                        }
                    }
                }
            }
        }
        reader.close()
    }
}